precompiles/assets/types.go

```diff
@@ -32,27 +35,132 @@ func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []i
+	if uint32(len(stakerAddr)) < clientChainAddrLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, len(stakerAddr), clientChainAddrLength)
 	}
 	depositWithdrawParams.StakerAddress = stakerAddr[:clientChainAddrLength]
 
 	opAmount, ok := args[3].(*big.Int)
 	if !ok || opAmount == nil || opAmount.Cmp(big.NewInt(0)) == 0 {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", opAmount)
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", args[3])
 	}
 	depositWithdrawParams.OpAmount = sdkmath.NewIntFromBigInt(opAmount)
 
 	return depositWithdrawParams, nil
 }
+
+func (p Precompile) ClientChainInfoFromInputs(_ sdk.Context, args []interface{}) (*types.ClientChainInfo, error) {
+	if len(args) != 5 {
+		return nil, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 5, len(args))
+	}
+	clientChain := types.ClientChainInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	clientChain.LayerZeroChainID = uint64(clientChainID)
+
+	addressLength, ok := args[1].(uint8)
+	if !ok || addressLength == 0 {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "uint8", args[1])
+	}
+	if addressLength < types.MinClientChainAddrLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, addressLength, types.MinClientChainAddrLength)
+	}
+	clientChain.AddressLength = uint32(addressLength)
+
+	name, ok := args[2].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, "string", args[2])
+	}
+	if name == "" || len(name) > types.MaxChainTokenNameLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidNameLength, name, len(name), types.MaxChainTokenNameLength)
+	}
+	clientChain.Name = name
+
+	metaInfo, ok := args[3].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "string", args[2])
+	}
+	if metaInfo == "" || len(metaInfo) > types.MaxChainTokenMetaInfoLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidMetaInfoLength, metaInfo, len(metaInfo), types.MaxChainTokenMetaInfoLength)
+	}
+	clientChain.MetaInfo = metaInfo
+
+	signatureType, ok := args[4].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 4, "string", args[4])
+	}
+	clientChain.SignatureType = signatureType
+
+	return &clientChain, nil
+}
+
+func (p Precompile) TokenFromInputs(ctx sdk.Context, args []interface{}) (types.AssetInfo, error) {
+	if len(args) != 6 {
+		return types.AssetInfo{}, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 6, len(args))
+	}
+	asset := types.AssetInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	asset.LayerZeroChainID = uint64(clientChainID)
+	info, err := p.assetsKeeper.GetClientChainInfoByIndex(ctx, asset.LayerZeroChainID)
```

Comment:
I see `p.assetsKeeper.GetClientChainInfoByIndex` would return error if client chain is not registered before, just comment to get your confirmation about this if my understanding is ture

Reply:
Yeah, the asset registration will depend on the related client chain's registration.